src/main/java/com/google/api/generator/engine/ast/MethodDefinition.java

```diff
@@ -277,15 +277,14 @@ public MethodDefinition build() {
         if (method.returnExpr() != null && !isLastStatementThrowExpr) {
           if (method.returnType().isPrimitiveType()) {
             Preconditions.checkState(
-                method.returnExpr().type().isPrimitiveType()
-                    && method.returnType().equals((method.returnExpr().type())),
-                "Method primitive return type does not match the return expression type");
+                method.returnType().equals((method.returnExpr().type())),
+                "Method primitive return type is not equal to the return expression type");
 
           } else {
             Preconditions.checkState(
-                !method.returnExpr().type().isPrimitiveType()
-                    && method.returnType().isSupertypeOrEquals(method.returnExpr().type()),
-                "Method reference return type is not a subtype of the return expression type");
+                method.returnType().equals(method.returnExpr().type())
+                    || method.returnType().isSupertypeOrEquals(method.returnExpr().type()),
+                "Method reference return type is not a super/boxed type of the return expression type");
```

Comment:
Cool, kept the original checks to make them clearer.

1. removed redundant `isPrimitiveType()` in the first block, since `returnExpr` type can be reference type for boxed/primitive type equality.
2. removed `isPrimitiveType()` check in else block, but `isSupertypeOrEquals` only checks when both types are reference types. So added `equals()` condition to allow boxed/primitive types.

Reply:
The `equals()` check is now partly redundant. Instead, let's make the first primitive-check condition `if (method.returnType().isPrimitiveType() || method.returnExpr().type().isPrimitiveType())` and keep the equals check there. 

That will also let us keep the original message here.

Reply:
Yes, the `equals` check in second block is only for boxed/primitive type equality which is partly redundant. That's a great idea, moved the primitive types check to the first block and reverted back the message change.